socketio/src/asynchronous/client/builder.rs

```diff
@@ -343,26 +385,70 @@ impl ClientBuilder {
     /// }
     /// ```
     pub async fn connect(self) -> Result<Client> {
+        let reconnect_delay_min = self.reconnect_delay_min;
+        let reconnect_delay_max = self.reconnect_delay_max;
+        let max_reconnect_attempts = self.max_reconnect_attempts;
+
         let socket = self.connect_manual().await?;
-        let socket_clone = socket.clone();
+        let mut socket_clone = socket.clone();
 
         // Use thread to consume items in iterator in order to call callbacks
         tokio::runtime::Handle::current().spawn(async move {
-            let mut stream = socket_clone.as_stream();
-            // Consume the stream until it returns None and the stream is closed.
-            while let Some(item) = stream.next().await {
-                if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = item {
-                    trace!("Network error occurred: {}", e.unwrap_err());
+            loop {
+                let mut stream = socket_clone.as_stream();
+                // Consume the stream until it returns None and the stream is closed.
+                while let Some(item) = stream.next().await {
+                    // Where does this get thrown by the async client? :confused:
+                    if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = item {
+                        trace!("Network error occurred: {}", e.unwrap_err());
+                    }
+                }
+
+                // Drop the stream so we can once again use `socket_clone` as mutable
+                drop(stream);
+
+                if socket_clone.should_reconnect().await {
+                    let mut reconnect_attempts = 0;
+                    let mut backoff = ExponentialBackoffBuilder::new()
+                        .with_initial_interval(Duration::from_millis(reconnect_delay_min))
+                        .with_max_interval(Duration::from_millis(reconnect_delay_max))
+                        .build();
+
+                    loop {
+                        if let Some(max_reconnect_attempts) = max_reconnect_attempts {
+                            reconnect_attempts += 1;
+                            if reconnect_attempts > max_reconnect_attempts {
+                                // TODO: How to tell the app about this?
```

Comment:
Logging is our best resort at the moment, since there is no way of telling this through a `Result` value at the moment.... Same as with the network error above.